app/services/samples/metadata/update_service.rb

```diff
@@ -0,0 +1,101 @@
+        @project = project
+        @sample = sample
+        @metadata = params['metadata']
+        @analysis_id = params['analysis_id']
+      end
+
+      def execute
+        authorize! sample.project, to: :update_sample?
+
+        validate_sample_in_project
+
+        validate_metadata_param
+
+        transform_metadata_keys
+
+        metadata_fields_update_status = metadata_modification
+
+        @sample.save
+
+        fields_not_updated(metadata_fields_update_status)
+        metadata_fields_update_status
+      rescue Samples::Metadata::UpdateService::SampleMetadataUpdateError => e
+        @sample.errors.add(:base, e.message)
+        metadata_fields_update_status || { updated: [], not_updated: [] }
+      end
+
+      private
+
+      def validate_sample_in_project
+        return unless @project.id != @sample.project.id
+
+        raise SampleMetadataUpdateError,
+              I18n.t('services.samples.metadata.sample_does_not_belong_to_project', sample_name: @sample.name,
+                                                                                    project_name: @project.name)
+      end
+
+      def validate_metadata_param
+        return unless @metadata.nil? || @metadata == {}
+
+        raise SampleMetadataUpdateError,
+              I18n.t('services.samples.metadata.empty_metadata', sample_name: @sample.name)
+      end
+
+      def transform_metadata_keys
+        # Without transforming keys, issues with overwritting can occur and multiples of the same key can appear
+        @metadata = @metadata.transform_keys(&:to_s)
+      end
+
+      def metadata_modification
+        update_status = { updated: [], not_updated: [] }
+        @metadata.each do |key, value|
+          if value.blank?
+            if @sample['metadata'].key?(key)
+              @sample['metadata'].delete(key)
+              @sample['metadata_provenance'].delete(key)
+              update_status[:updated].append(key)
+            end
+          else
+            metadata_updated = assign_metadata_to_sample(key, value)
+            metadata_updated ? update_status[:updated].append(key) : update_status[:not_updated].append(key)
+          end
+        end
+        update_status
+      end
+
+      def assign_metadata_to_sample(key, value)
+        # We don't overwrite existing @sample['metadata_provenance'] or @sample['metadata']
+        # that has a {source: 'analysis'} with a user
+        if @sample['metadata_provenance'].key?(key) && @analysis_id.nil? &&
+           @sample['metadata_provenance'][key]['source'] == 'analysis'
+          false
+        else
+          @sample['metadata_provenance'][key] =
+            @analysis_id.nil? ? { source: 'user', id: current_user.id } : { source: 'analysis', id: @analysis_id }
+          @sample['metadata'][key] = value
+          true
+        end
+      end
+
+      def fields_not_updated(metadata_fields_update_status)
+        return unless metadata_fields_update_status.count.positive?
```

Comment:
Since we are only interested in the `not_updated` fields, assign `metadata_fields_update_status[:not_updated]` to a variable then check the count on that, and then use the variable below when setting the metadata_fields in the error

Reply:
Updated in 5ab62ea9e785867b0a42761d594ba86b2f4f0a82